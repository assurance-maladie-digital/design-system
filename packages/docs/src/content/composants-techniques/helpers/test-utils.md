---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
