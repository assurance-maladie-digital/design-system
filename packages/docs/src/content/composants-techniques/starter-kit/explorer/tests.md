Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
