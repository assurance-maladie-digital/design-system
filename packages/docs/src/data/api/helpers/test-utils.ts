import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
