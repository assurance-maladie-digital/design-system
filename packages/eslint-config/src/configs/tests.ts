import { type FlatESLintConfigItem } from 'eslint-define-config';
import { pluginVitest } from '../plugins';
import { GLOB_TESTS } from '../globs';
import { isInEditor } from '../env';

export const tests: FlatESLintConfigItem[] = [
	{
		plugins: {
			tests: pluginVitest,
		},
	},
	{
		files: GLOB_TESTS,
		rules: {
			'tests/consistent-test-it': [
				'error',
				{
					fn: 'it',
					withinDescribe: 'it',
				},
			],
			'tests/expect-expect': 'error',
			'tests/no-alias-methods': 'error',
			'tests/no-commented-out-tests': isInEditor ? 'off' : 'error',
			'tests/no-conditional-expect': 'error',
			'tests/no-disabled-tests': 'error',
			'tests/no-done-callback': 'error',
			'tests/no-duplicate-hooks': 'error',
			'tests/no-focused-tests': isInEditor ? 'off' : 'error',
			'tests/no-identical-title': 'error',
			'tests/no-interpolation-in-snapshots': 'error',
			'tests/no-standalone-expect': 'error',
			'tests/no-test-return-statement': 'error',
			'tests/prefer-comparison-matcher': 'error',
			'tests/prefer-equality-matcher': 'error',
			'tests/prefer-hooks-in-order': 'error',
			'tests/prefer-hooks-on-top': 'error',
			'tests/prefer-mock-promise-shorthand': 'error',
			'tests/prefer-spy-on': 'error',
			'tests/prefer-to-be-falsy': 'error',
			'tests/prefer-to-be-object': 'error',
			'tests/prefer-to-be-truthy': 'error',
			'tests/prefer-to-be': 'error',
			'tests/prefer-to-contain': 'error',
			'tests/prefer-to-have-length': 'error',
			'tests/prefer-todo': 'error',
			'tests/require-hook': 'error',
			'tests/require-top-level-describe': 'error',
			'tests/valid-describe-callback': 'error',
			'tests/valid-expect': 'error',
			'tests/valid-title': 'error',
		},
	},
];
