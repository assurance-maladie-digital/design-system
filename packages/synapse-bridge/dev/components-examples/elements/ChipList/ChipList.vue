<template>
	<ChipList :items="items" />
	<ChipList
		:items="items"
		:overflow-limit="10"
	/>
</template>

<script lang="ts">
import ChipList from "@/elements/ChipList";
import {ChipItem} from "@/elements/ChipList/types";

export default {
	components: {
		ChipList,
	},
	data() {
		return {
			items: [
				{ text: "Item 1" },
				{ text: "Item 2" },
				{ text: "Item 3" },
				{ text: "Item 4" },
				{ text: "Item 5" },
				{ text: "Item 6" },
				{ text: "Item 7" },
				{ text: "Item 8" },
				{ text: "Item 9" },
				{ text: "Item 10" },
				{ text: "Item 11" },
				{ text: "Item 12" },
				{ text: "Item 13" },
				{ text: "Item 14" },
				{ text: "Item 15" },
				{ text: "Item 16" },
				{ text: "Item 17" },
				{ text: "Item 18" },
				{ text: "Item 19" },
				{ text: "Item 20" },
			] as ChipItem[],
		};
	},
};
</script>
