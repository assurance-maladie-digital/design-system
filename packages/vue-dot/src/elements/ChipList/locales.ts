export const locales = {
	closeBtnLabel: 'Supprimer le filtre pour cette valeur',
	reset: 'RÃ©initialiser le filtre'
};
