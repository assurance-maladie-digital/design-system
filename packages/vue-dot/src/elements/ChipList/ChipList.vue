<template>
	<div
		v-if="items.length"
		:class="{
			'flex-column': showOverflowChip
		}"
		class="vd-chip-list d-flex flex-wrap max-width-none mx-n1 mt-n1"
	>
		<div class="d-flex flex-wrap align-center">
			<VChip
				v-for="(item, index) in filteredItems"
				:key="index"
				:close-icon="deleteIcon"
				:close-label="locales.closeBtnLabel"
				color="cyan-darken-40"
				text-color="white"
				close
				small
				class="ma-1"
				@click:close="emitRemoveEvent(item)"
			>
				{{ item.text }}
			</VChip>
		</div>

		<div class="ml-1">
			<VChip
				v-if="showOverflowChip"
				color="cyan-lighten-90"
				text-color="cyan-darken-40"
				small
				class="vd-overflow-chip ma-1"
			>
				{{ overflowText }}
			</VChip>

			<VBtn
				color="primary"
				small
				text
				class="vd-overflow-btn px-1 ml-0 my-1"
				@click="emitResetEvent"
			>
				{{ locales.reset }}
			</VBtn>
		</div>
	</div>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component from 'vue-class-component';

	import { locales } from './locales';
	import { ChipItem } from './types';

	import { mdiChevronUp, mdiWindowClose } from '@mdi/js';

	const Props = Vue.extend({
		props: {
			items: {
				type: Array as PropType<ChipItem[]>,
				default: () => []
			},
			overflowLimit: {
				type: Number,
				default: 4
			}
		}
	});

	@Component
	export default class ChipList extends Props {
		locales = locales;

		deleteIcon = mdiWindowClose;
		upIcon = mdiChevronUp;

		get showOverflowChip(): boolean {
			return this.items.length >= this.overflowLimit;
		}

		get overflowText(): string {
			return `+${this.items.length - this.overflowLimit + 1}`;
		}

		get filteredItems(): ChipItem[] {
			return this.items.slice(0, this.overflowLimit - 1);
		}

		emitRemoveEvent(item: ChipItem): void {
			this.$emit('remove', item);
		}

		emitResetEvent(): void {
			this.$emit('reset');
		}
	}
</script>

<style lang="scss" scoped>
	@import '@cnamts/design-tokens/dist/tokens';

	.vd-overflow-chip {
		border: 1px solid $vd-cyan-lighten-60 !important;
	}

	.vd-overflow-btn {
		// Disable hover state
		&.theme--light::before {
			content: none;
		}

		&:focus-visible {
			outline: 2px solid;
		}
	}
</style>
