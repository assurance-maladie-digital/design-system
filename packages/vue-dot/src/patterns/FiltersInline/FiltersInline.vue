<template>
	<div class="vd-filters-inline d-flex flex-wrap max-width-none ma-n2">
		<VMenu
			v-for="filter in filters"
			:key="filter.name"
			:close-on-content-click="false"
			min-width="300px"
			max-width="400px"
			offset-y
			nudge-bottom="10px"
		>
			<template #activator="{ on }">
				<VBtn
					:class="`vd-filter-${filter.name}`"
					:outlined="!getFilterCount(filter)"
					color="secondary"
					depressed
					rounded
					class="ma-2"
					v-on="on"
				>
					<VBadge
						v-if="getFilterCount(filter)"
						:content="getFilterCount(filter)"
						:label="locales.badgeLabel(getFilterCount(filter))"
						color="white"
						inline
						class="ml-n1 mr-1"
					/>

					{{ filter.title }}

					<VIcon
						small
						class="ml-1"
					>
						{{ downIcon }}
					</VIcon>
				</VBtn>
			</template>

			<VSheet
				background="white"
				class="pa-4"
			>
				<ChipList
					:items="getChips(filter)"
					:overflow-limit="filter.chipOverflowLimit"
					class="mb-5"
					@remove="removeChip(filter, $event)"
					@reset="resetFilter(filter)"
				/>

				<slot
					:on="{
						input: value => {
							$set(filter, 'value', value);
							updateValue();
						},
						change: value => {
							$set(filter, 'value', value);
							updateValue();
						}
					}"
					:attrs="{
						value: filter.value
					}"
					:name="`${formatFilterName(filter.name)}`"
				/>
			</VSheet>
		</VMenu>
	</div>
</template>

<script lang="ts">
	import Component from 'vue-class-component';

	import { locales } from './locales';

	import { Filterable } from '../../mixins/filterable';

	import { mdiChevronDown } from '@mdi/js';

	@Component
	export default class FiltersInline extends Filterable {
		locales = locales;

		downIcon = mdiChevronDown;
	}
</script>

<style lang="scss" scoped>
	@import '@cnamts/design-tokens/dist/tokens';

	.v-badge :deep(.v-badge__badge) {
		color: $vd-secondary !important;
	}
</style>
