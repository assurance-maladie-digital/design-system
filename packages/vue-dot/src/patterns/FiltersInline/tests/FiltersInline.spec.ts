import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';

import FiltersInline from '../';
import ChipList from '../../../elements/ChipList';

Vue.component('ChipList', ChipList);

let wrapper: Wrapper<Vue>;

describe('FiltersInline', () => {
	it('renders correctly', () => {
		wrapper = mountComponent(FiltersInline, {
			propsData: {
				value: [
					{
						name: 'name',
						label: 'Nom'
					}
				]
			}
		});

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with an active filter', () => {
		wrapper = mountComponent(FiltersInline, {
			propsData: {
				value: [
					{
						name: 'name',
						label: 'Nom',
						value: 'John Doe'
					}
				]
			}
		}, true);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with multiple active filters', () => {
		wrapper = mountComponent(FiltersInline, {
			propsData: {
				value: [
					{
						name: 'name',
						label: 'Nom',
						value: 'John Doe'
					},
					{
						name: 'profession',
						label: 'Profession',
						value: [
							'Infirmer',
							'Pharmacien'
						]
					}
				]
			}
		}, true);

		expect(wrapper).toMatchSnapshot();
	});
});
