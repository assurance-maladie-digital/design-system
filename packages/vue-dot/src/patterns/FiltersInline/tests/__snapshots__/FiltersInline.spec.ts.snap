// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersInline renders correctly 1`] = `
<div class="vd-filters-inline d-flex flex-wrap max-width-none ma-n2">
  <vmenu-stub opendelay="0" closedelay="0" openonclick="true" contentclass="" maxwidth="400px" minwidth="300px" nudgebottom="10px" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" closeonclick="true" maxheight="auto" offsety="true" origin="top left" transition="v-menu-transition" contentprops="[object Object]">
    <vsheet-stub tag="div" background="white" class="pa-4">
      <chiplist-stub items="" overflowlimit="4" class="mb-5"></chiplist-stub>
    </vsheet-stub>
  </vmenu-stub>
</div>
`;

exports[`FiltersInline renders correctly with an active filter 1`] = `
<div class="vd-filters-inline d-flex flex-wrap max-width-none ma-n2">
  <div class="v-menu"><button type="button" class="ma-2 v-btn v-btn--has-bg v-btn--rounded theme--light v-size--default secondary vd-filter-name"><span class="v-btn__content"><span class="v-badge ml-n1 mr-1 v-badge--inline theme--light"><span class="v-badge__wrapper"><transition-stub name="scale-rotate-transition"><span aria-atomic="true" aria-label="1 filtre" aria-live="polite" role="status" class="v-badge__badge white">1</span></transition-stub></span></span>



      <span aria-hidden="true" class="v-icon notranslate ml-1 theme--light" style="font-size: 16px; height: 16px; width: 16px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg" style="font-size: 16px; height: 16px; width: 16px;"><path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path></svg></span></span></button>
    <!---->
  </div>
</div>
`;

exports[`FiltersInline renders correctly with multiple active filters 1`] = `
<div class="vd-filters-inline d-flex flex-wrap max-width-none ma-n2">
  <div class="v-menu"><button type="button" class="ma-2 v-btn v-btn--has-bg v-btn--rounded theme--light v-size--default secondary vd-filter-name"><span class="v-btn__content"><span class="v-badge ml-n1 mr-1 v-badge--inline theme--light"><span class="v-badge__wrapper"><transition-stub name="scale-rotate-transition"><span aria-atomic="true" aria-label="1 filtre" aria-live="polite" role="status" class="v-badge__badge white">1</span></transition-stub></span></span>



      <span aria-hidden="true" class="v-icon notranslate ml-1 theme--light" style="font-size: 16px; height: 16px; width: 16px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg" style="font-size: 16px; height: 16px; width: 16px;"><path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path></svg></span></span></button>
    <!---->
  </div>
  <div class="v-menu"><button type="button" class="ma-2 v-btn v-btn--has-bg v-btn--rounded theme--light v-size--default secondary vd-filter-profession"><span class="v-btn__content"><span class="v-badge ml-n1 mr-1 v-badge--inline theme--light"><span class="v-badge__wrapper"><transition-stub name="scale-rotate-transition"><span aria-atomic="true" aria-label="2 filtres" aria-live="polite" role="status" class="v-badge__badge white">2</span></transition-stub></span></span>



      <span aria-hidden="true" class="v-icon notranslate ml-1 theme--light" style="font-size: 16px; height: 16px; width: 16px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg" style="font-size: 16px; height: 16px; width: 16px;"><path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path></svg></span></span></button>
    <!---->
  </div>
</div>
`;
